### ISSUE TITLE
android support

### ISSUE THREAD

**Issue by RumenTech** [rogchap/react-native-twilio#3]
Thanks for your nice wrapper, I just wanted to know when it comes to support android...

**Comment by rogchap** [comment]
I'm currently working on building out an iOS app. 
Once I get feedback on the iOS version that will determine the timeline for Android.

Honestly, although this is on my todo list, it may take some time to get done.

Happy for contributions / PR requests. 

... might see if I can get something working this afternoon, which might be a good point for someone to finish off.

**Comment by rogchap** [comment]
0e740e5 I've added the basic outlines of the android version. 
There is no functinality, but just basic setup.
Includes the twilio android client so should be simple to follow the RN module guide to finish off:
http://facebook.github.io/react-native/docs/native-modules-android.html

**Comment by EugeneLiang** [comment]
Just wondering if there's any news on Android support ?

**Comment by rogchap** [comment]
I've started the implementation: https://github.com/rogchap/react-native-twilio/commit/4d3d980bb182b2fdf291e095b9b6c32a1a9c639f

Can anyone help me with testing? As the code is 100% un-tested!

**Comment by EugeneLiang** [comment]
@rogchap By testing do you mean just using the code or write unit tests ? Cheers.

**Comment by rogchap** [comment]
@EugeneLiang just use it and see if it 1. Works 2.throws any errors. 

I don't have time to put a full android example app together at the moment.

Please use the develop branch.

**Comment by EugeneLiang** [comment]
@rogchap  ok cool. i'll volunteer as a tribute.

**Comment by fabriziomoscon** [comment]
@rogchap I can see the `android` skeleton code in `master` but the code is not on `npm`. I am planning to use it and contribute if I can. Could you please publish the `master` version on `npm`?

**Comment by fabriziomoscon** [comment]
I started the Android implementation:
https://github.com/fabriziomoscon/react-native-twilio/tree/develop

work in progress

**Comment by rogchap** [comment]
@fabriziomoscon have you tested the develop branch? The Android implementation is already there.

**Comment by fabriziomoscon** [comment]
yes definitely, (I should have said... I am continuing the Android implementation)
my work is based on the `develop` branch. When I first looked at the code, it didn't even compile. I am building it along my phone app... long process but I have already few good results.

If you have experience with Android I could definitely do with some help

**Comment by rogchap** [comment]
Thanks @fabriziomoscon 

No surprise it did not compile... I didn't even try; just put together what I thought was needed.

**Comment by jswilson** [comment]
I've been working on this as well:

https://github.com/jswilson/react-native-twilio/commit/4506ed5169f834d62b7821a342acdf5a533e8240

I have run into a major problem _accepting_ calls, which is my primary need on Android.  With this code, I can successfully accept a call, and it works as expected.  But when the call is disconnected, I get a crash when trying to send the event over the bridge to javascript-world.  And it's interesting: after Twilio disconnects, any dereferencing at all of the DeviceEventEmitter javascript module causes a crash, not just emitting a message.  I actually watch for the disconnect event, and just call toString() on the module instead of emitting the event, and I still get a crash (I've also checked the reference, and it is not null).  Would you guys have any thoughts as to what might be going wrong here (or maybe the right forum to ask about this)?  I have a test app that uses this code, if that would be helpful.  Note, I do need to stay on the call for about 1.5 minutes to see this behavior.

Thank you both for your work in getting things going on Android!

**Comment by fabriziomoscon** [comment]
Good work on the receiving part! I haven't tested it yet as I want to complete the making call. Please see my fork as well to see what am I doing right and wrong. Together we will be faster to produce this.

I filed an issue here: https://github.com/twilio/twilio-client-quickstart-android/issues/13
not sure if that is related though.

I got a feeling that we should get the folks of Twilio involved to speed up the debugging

**Comment by fabriziomoscon** [comment]
@jswilson I looked at your code and used part of it to implement the call receiving methods.
Have a look at this:
https://github.com/fabriziomoscon/react-native-twilio/commit/846205a9962f969c93f2eb0c553777442e6ba95b

I discovered that Twilio lib uses a Service intent to communicate when a call is received. And that works both with the app in foreground and background.

@jswilson could you share the way you use the module in the app? For example the relevant section of your `build.gradle`?
Also I notice that you accept all incoming calls immediately
https://github.com/jswilson/react-native-twilio/blob/develop/android/src/main/java/com/rogchap/react/modules/twilio/TwilioModule.java#L53
is that part of your use case? I would expect the user to accept the call from the UI

UPDATE
I had to swap these 2 lines to prevent the crash on receiving the call:
https://github.com/jswilson/react-native-twilio/blob/develop/android/src/main/java/com/rogchap/react/modules/twilio/TwilioModule.java#L52-L53

**Comment by jswilson** [comment]
@fabriziomoscon  I'm happy to share how I'm using the module in the app.  I don't think there's anything exotic in my build.gradle...the only special thing I can think of is adding the Service to the manifest file.  Basically, I added this line to the `<application>` section of AndroidManifest.xml:

```
  <service android:enabled="true" android:name="com.twilio.client.TwilioClientService" />
```

I'll work on putting together a sample app that uses the lib and I'll publish that.

The auto-accept was a shortcut I took, you are absolutely correct that accept shouldn't be called upon receipt; that would actually be quite bad to do in production, because I believe you can receive an incoming call while a call is in progress (that would need verification, though).  My use case is pretty limited, though, so I just automatically accepted them as I was testing things out.

Looks like I uploaded a slightly out-of-date version; you're right, it does crash without swapping those lines.

I put this issue down for a little while, so I don't have much new to report, but I am hoping to tackle it a bit this week.

**Comment by fabriziomoscon** [comment]
Great to hear from you. Yes I am also using `<service ...` in the Manifest.

I have started putting together a react-native example here: https://github.com/fabriziomoscon/react-native-twilio/tree/develop/RNTwilioExample
so I can isolate the bug without the rest of the code... it would be great if you can have a look and tell me what you think. Maybe we can both work on that one so to avoid rewriting a new sample app ;)

I have also asked some help to Twilio via this repo: https://github.com/twilio/twilio-client-quickstart-android/issues/13

Please let me know about anything you find

**Comment by fabriziomoscon** [comment]
Update:
I managed to fix few crashes in the android module. I need some more work, but at this point it looks very promising.

Question for @jswilson when you receive a call and the caller hang up before the device answer I can't see any event triggered. Did you solve that case?

**Comment by rogchap** [comment]
Nice one @fabriziomoscon looking forward to seeing the end result.

**Comment by Pagebakers** [comment]
I'll try to get it to run later this week and see what happens, thanks for this!

**Comment by jswilson** [comment]
@fabriziomoscon Sorry I've been so out of the loop.  I actually have not solved that case.  My work is mostly at a prototype stage, so I haven't been too worried about it, but it is definitely something on my radar.  Is the complete callback fired server-side in that case?  See [here](https://www.twilio.com/blog/2015/05/introducing-call-progress-events-flexibly-track-and-control-your-outbound-calls.html) for the callbacks provided.  

The work you have done on the Android side is fantastic.  I believe all of my crash problems centered around the onStartListening crash you describe in the comments.  I spent more time debugging that than I care to admit!  Awesome find, I am in your debt!  

Are there other problems you are actively debugging?  For what it’s worth, I’ve been testing out your library of most of today and it has been working great.  My use case is limited to receiving only on Android, so take that for what it's worth, but it is definitely working great for me.  I've got a minor fix I'll push your way regarding initWithTokenUrl.

**Comment by fabriziomoscon** [comment]
@jswilson I am looking into callback events right now. That might be the best way to receive the proper feedback event for call completed even when the caller hangs up, but the round trip or alerting the client from the server might require some expensive websocket or some server push equivalent idea. I will hunt down this issue when the other crashes are fixed.

All the remaining crashes seems to be rooted in sending events to javascript. It might depend on the way Java threads work, or/and the way Twilio code call the @Overwritten methods of the listeners and I not having much knowledge about those. I think I am doing something wrong in there.

I have opened an issue on reaact-native to see if the guys at Facebook can shed some lights: https://github.com/facebook/react-native/issues/9379

Basically the app crashes when the second event is being prepared to be sent to Javascript. The crash is inside the react-native code.

**Comment by jswilson** [comment]
Interesting...my Android app is receive-only, and I'm now not getting any crashes on it.  I've made a bunch of calls of varying lengths and under different conditions, and I haven't seen one crash yet.  I did put together an Android app that initiates calls, and I do see the crash there.  Not sure if that helps, or if you're having the same experience, but that is what I'm currently seeing.

**Comment by fabriziomoscon** [comment]
did you try to use this example app? https://github.com/fabriziomoscon/react-native-twilio/tree/develop/RNTwilioExample
That is very interesting. What I see is that receiving the second disconnecting event: https://github.com/fabriziomoscon/react-native-twilio/blob/develop/android/src/main/java/com/rogchap/react/modules/twilio/TwilioModule.java#L314-L340
I have a crash.
Out of curiosity are you receiving calls also using the background service or jut when the app is open and on the foreground?

**Comment by jswilson** [comment]
I haven't tried the example app yet, I will give definitely give that a shot to see if I get the same results.

I am only receiving calls when the app is open and in the foreground.

**Comment by fabriziomoscon** [comment]
I was looking at this https://github.com/ccm-innovation/react-native-twilio-ip-messaging and it looks like that project has implemented the sending events to JavaScript in a different way than I was doing here:
https://github.com/fabriziomoscon/react-native-twilio/blob/merge-upstream/android/src/main/java/com/rogchap/react/modules/twilio/TwilioModule.java#L98

@bradbumbalough could I ask if you have an idea why my implementation would crash at the second time an event is sent?

**Comment by bradbyte** [comment]
@fabriziomoscon it's hard to say for certain without digging deep into your code, as it looks like we are doing a similar event implementation? unless you changed it recently?

My approach was basically a singleton pattern for the React classes (same with my iOS side). I believe that the React bridge will create new instances of the classes, which is why I wanted things to persist globally. 

Are you sure your params on the 2nd call are legit? I remember having issues with "crashes" because of bad or dirty params, not necessarily the interface. They were getting caught in the native bridge code too. It might be worth testing if you can successfully make a 2nd call with forced params, just to rule it out.

Best of luck!

**Comment by fabriziomoscon** [comment]
I have submitted a PR https://github.com/rogchap/react-native-twilio/pull/19
with all the relevant code for Android support + an Example project.

It is working for most of the cases. Although I still get some crash due to RN events here and then.

NOTE:
Twilio has moved to Voice Programmable SDK for native app VOIP, which basically uses push notifications.
https://www.twilio.com/docs/api/client/android

**Comment by rogchap** [comment]
Merged and published as **1.0.0**
Good work @fabriziomoscon ... ledgend.
